### ISSUE TITLE
Getting 504: Gateway Timeout while ingesting data

### ISSUE THREAD

**Issue by abhimech001** [vespa-engine/vespa#17869]
Hi Team,

I have vespa multinode cluster set up and I tried running ingestion load test on it with 80 users and payload size of 5MB.
1 config server
9 container server (CPU 1 core-6 core, Memory 2G-8G)
6 content node (CPU 1 core-2core, Memory 32G-32G)

I observed more than 50% requests failed with response code 504, response message "gateway timeout" and response message "upstream request timeout". 

1) How can we reduce the error rate?
2) What is the default request timeout value and threads for ingestion(seems to be 15s).Is there a way to increase ingestion timeout (like we do for search by adding timeout in http url) when using http request?

**Comment by abhimech001** [comment]
I have tried below url to increase timeout but the above error is consistent as soon as the request breaches 15s
http://<host url>/document/v1/<namespace>/<document-type>/docid/<document-id>?timeout=60s
POST
{"fields:" {"album": "value1", "artist": "value2","year": "value3"}}

**Comment by bjormel** [comment]
Hi,

In the [Feed Sizing Guide](https://docs.vespa.ai/en/performance/sizing-feeding.html), there is a Troubleshooting section which references the [Metrics](https://docs.vespa.ai/en/reference/metrics.html) documentation.

When testing feed throughput you should gradually increase the load until you start to see the 504s, then you can inspect the metrics to try to identify the bottleneck.

**Comment by bratseth** [comment]
Yes, you can set timeout, see https://docs.vespa.ai/en/reference/document-v1-api-reference.html
Note that if you want to feed as fast as possible (but no faster) you should use the https://docs.vespa.ai/en/vespa-http-client.html
In any case, it is a good idea to graph enough metrics to know where you are running out of capacity.

**Comment by jobergum** [comment]
> payload size of 5MB.

So average document size is 5MB per document? mainly text or?  

> 9 container server (CPU 1 core-6 core, Memory 2G-8G)

You mean different amount of memory available per node when you write 2G-8G? You might run into jvm memory issues on container nodes with just 2G. With 5MB documents with some level of concurrency I expect that this will cause high memory. I would throw out nodes with less than 8G, and adjust heap size to 4G, just like the example here https://docs.vespa.ai/en/performance/container-tuning.html#jvm-tuning

**Comment by abhimech001** [comment]
> t amount of memory available per node when you write 2G-8G? You might run into jvm memory issues on container nodes wi


Thanks for the reply!!!

Kubernetes has a feature to provide the resource in burst mode. 2G to 8G means, it will start with 2G and can go max upto 8G on load and when the load subsides it reduces from 8G to 2G. During the load test I did monitor using top command and the pods were able to get the resources when requiered.
At present for 5MB payload size, 80 threads ,30 sec client timeout and 30 sec request timeout - 25% requests fail with gateway timeout. Reducing the timeout results in further increase in error rate.
I will try take look at metrics and understand the json and revert if I need anything.

**Comment by jobergum** [comment]
That dynamic memory pod changes is not something that the vespa document processing pipeline will benefit from. The indexing document processor is within a Java container which startup parameters determines heap size settings. If you run low on heap you might run into garbage collection pressure (Which is likely with 5MB documents flying through with concurrency 80 = 4G of inflight data).

**Comment by abhimech001** [comment]
> That dynamic memory pod changes is not something that the vespa document processing pipeline will benefit from. The indexing document processor is within a Java container which startup parameters determines heap size settings. If you run low on heap you might run into garbage collection pressure (Which is likely with 5MB documents flying through with concurrency 80 = 4G of inflight data).

Thanks for the reply
I will try by updating jvm to 4G and revert post my testing

**Comment by abhimech001** [comment]
> That dynamic memory pod changes is not something that the vespa document processing pipeline will benefit from. The indexing document processor is within a Java container which startup parameters determines heap size settings. If you run low on heap you might run into garbage collection pressure (Which is likely with 5MB documents flying through with concurrency 80 = 4G of inflight data).

Hi,

Changed stateless pod memory (9) to 8G-8G and tuned jvm to 50% of allocated memory via services.xml.
Still the error rate continued to be more than 25%.

**Comment by jobergum** [comment]
Reduce concurrency then or use the async Vespa http client library which
solves this.

**Comment by jobergum** [comment]
Do you have a proxy (e.g nginx) in front of the containers? 

>upstream request timeout

Is not something that I would expect vespa containers to write. Also the default vespa feed timeout is 180 seconds, and not 15 seconds.

**Comment by abhimech001** [comment]
> Do you have a proxy (e.g nginx) in front of the containers?
> 
> > upstream request timeout
> 
> Is not something that I would expect vespa containers to write. Also the default vespa feed timeout is 180 seconds, and not 15 seconds.

Thanks for the suggestion. It helped. Increased proxy output to 180s and ran load test. error rate reduced to acceptable limits.

**Comment by abhimech001** [comment]
> I'm resolving this. I think you need a bit more modern HW to fully enjoy Vespa. Frequent latency spikes above 15 seconds are signs of an overloaded system.

Thank you!!!

One last ask here. We are using NAS storage(data persistence) and can see degraded perf. Can you please help name some modern HW to enjoy vespa.

**Comment by jobergum** [comment]
Quick summary

- Local  fast (SSD) storage is preferred. 
- Writes to transaction log is per default synched, this might have a severe performance penalty on NAS storage. See https://docs.vespa.ai/en/reference/services-content.html#sync-transactionlog
- Content nodes compute type of nodes with local ssd and memory for your document types, for example on AWS Ec2: c5d.9xlarge would be a good fit for many use cases. 36 v-cpu, 72G memory and 1 900 NVMe SSD. This is high end though and obviously use case dependent (how many qps, latency requirements etc) 

Also our hosted Vespa offering cloud.vespa.ai should be mentioned as a great alternative
